plugins/hls-refine-imports-plugin/src/Ide/Plugin/RefineImports.hs

```diff
@@ -0,0 +1,293 @@
+        let importedNames = map (prettyPrint . ieName . unLoc) names
+        in flip Map.filter avails $ \a ->
+              any ((`elem` importedNames) . prettyPrint) 
+                $ concatMap availNamesWithSelectors a
+      filterByImport _ _ = mempty
+  let constructImport 
+        :: LImportDecl GhcRn 
+        -> (ModuleName, [AvailInfo]) 
+        -> LImportDecl GhcRn
+      constructImport 
+        i@(L lim id@ImportDecl
+                  {ideclName = L _ mn, ideclHiding = Just (hiding, L _ names)})
+        (newModuleName, avails) = L lim id
+          { ideclName = noLoc newModuleName
+          , ideclHiding = Just (hiding, noLoc newNames)
+          }
+          where newNames = filter (\n -> any (n `containsAvail`) avails) names
+      constructImport lim _ = lim
+  let res =
+        [ (i, Just
+                . T.intercalate "\n" 
+                . map (T.pack . prettyPrint . constructImport i) 
+                . Map.toList 
+                $ filteredInnerImports)
+        -- for every minimal imports
+        | Just minImports <- [mbMinImports]
+        , i@(L _ ImportDecl{ideclName = L _ mn}) <- minImports
+        -- we check for the inner imports
+        , Just innerImports <- [Map.lookup mn import2Map]
+        -- and only get those symbols used
+        , filteredInnerImports <- [filterByImport i innerImports]
+        -- if no symbols from this modules then don't need to generate new import
+        , not $ null filteredInnerImports
+        ]
+  return ([], RefineImportsResult res <$ mbMinImports)
+
+  where
+    -- Check if a name is exposed by AvailInfo (the available information of a module)
+    containsAvail :: LIE GhcRn -> AvailInfo -> Bool
+    containsAvail name avail = 
+      any (\an -> prettyPrint an == (prettyPrint . ieName . unLoc $ name)) 
+        $ availNamesWithSelectors avail
+
+--------------------------------------------------------------------------------
+
+-- | Use the ghc api to extract a minimal, explicit set of imports for this module
+extractMinimalImports ::
+  Maybe HscEnvEq ->
+  Maybe TcModuleResult ->
+  IO ([LImportDecl GhcRn], Maybe [LImportDecl GhcRn])
+extractMinimalImports (Just hsc) (Just TcModuleResult {..}) = do
+  -- extract the original imports and the typechecking environment
+  let tcEnv = tmrTypechecked
+      (_, imports, _, _) = tmrRenamed
+      ParsedModule {pm_parsed_source = L loc _} = tmrParsed
+      span = fromMaybe (error "expected real") $ realSpan loc
+
+  -- GHC is secretly full of mutable state
+  gblElts <- readIORef (tcg_used_gres tcEnv)
+
+  -- call findImportUsage does exactly what we need
+  -- GHC is full of treats like this
+  let usage = findImportUsage imports gblElts
+  (_, minimalImports) <- 
+    initTcWithGbl (hscEnv hsc) tcEnv span $ getMinimalImports usage
+
+  -- return both the original imports and the computed minimal ones
+  return (imports, minimalImports)
+extractMinimalImports _ _ = return ([], Nothing)
```

Comment:
It looks like there are lots of duplications between this plugin and `hls-explicit-imports-plugin`. Why don't we just put these function into that plugin?

Reply:
I exposed functions from explicit-imports and use them for those are exactly the same